x/marker/types/authz.go

```diff
@@ -99,3 +109,155 @@ func toBech32Addresses(allowed []sdk.AccAddress) []string {
+	}
+
+	anyAuths := make([]*codectypes.Any, len(subAuthorizations))
+	for i, auth := range subAuthorizations {
+		if auth == nil {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("sub-authorization %d is nil", i)
+		}
+		if auth.MsgTypeURL() != msgTypeURL {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("sub-authorization %d msg type mismatch", i)
+		}
+		if err := auth.ValidateBasic(); err != nil {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("sub-authorization %d failed basic validation: %v", i, err)
+		}
+		if _, isMulti := auth.(*MultiAuthorization); isMulti {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("nested MultiAuthorization not allowed for sub-authorization %d", i)
+		}
+
+		authValue, err := codectypes.NewAnyWithValue(auth)
+		if err != nil {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("failed to pack sub-authorization %d: %s", i, err)
+		}
+		anyAuths[i] = authValue
+	}
+
+	return &MultiAuthorization{
+		MsgTypeUrl:        msgTypeURL,
+		SubAuthorizations: anyAuths,
+	}, nil
+}
+
+// MsgTypeURL implements Authorization.MsgTypeURL.
+func (m MultiAuthorization) MsgTypeURL() string {
+	return m.MsgTypeUrl
+}
+
+// Accept implements Authorization.Accept.
+func (m MultiAuthorization) Accept(ctx context.Context, msg sdk.Msg) (authz.AcceptResponse, error) {
+	if m.MsgTypeURL() != sdk.MsgTypeURL(msg) {
+		return authz.AcceptResponse{}, sdkerrors.ErrInvalidType.Wrap("message type mismatch")
+	}
+
+	anyDeleteRequested := false
+	updatedAuths := make([]*codectypes.Any, len(m.SubAuthorizations))
+	anyUpdates := false
+
+	for i, anyAuth := range m.SubAuthorizations {
+		if anyAuth == nil {
+			return authz.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf("sub-authorization %d is nil", i)
+		}
+
+		auth, ok := anyAuth.GetCachedValue().(authz.Authorization)
+		if !ok || auth == nil {
+			return authz.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf("sub-authorization %d not unpacked", i)
+		}
+		resp, err := auth.Accept(ctx, msg)
+		if err != nil {
+			return authz.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf("sub-authorization %s failed", err)
+		}
+		if !resp.Accept {
+			return authz.AcceptResponse{Accept: false}, nil
+		}
+		// If any sub-authorization requests delete, mark for full deletion
+		if resp.Delete {
+			anyDeleteRequested = true
+		}
+		if resp.Updated != nil {
+			// Validate updated authorization
+			if resp.Updated.MsgTypeURL() != m.MsgTypeUrl {
+				return authz.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf(
+					"sub-authorization %d updated to wrong message type: expected %s, got %s",
+					i, m.MsgTypeUrl, resp.Updated.MsgTypeURL(),
+				)
+			}
+			if _, isMulti := resp.Updated.(*MultiAuthorization); isMulti {
+				return authz.AcceptResponse{}, sdkerrors.ErrInvalidRequest.Wrapf(
+					"nested MultiAuthorization not allowed in update for sub-authorization %d", i,
+				)
+			}
```

Comment:
I think these checks can be removed.

It's safe to assume the msg type url won't change during an `Accept` because a lot of other things would break if that were possible.

Similarly, it's safe to assume that the `resp.Updated` authorization is the same type as the one that `Accept` was called on. Since we make sure there aren't any nested `MultiAuthorization`s before writing the authorization to state, there's no way it'd end up with one in here.